### PR TITLE
[crypto] Reduce wasted pseudorandom bytes in ChaCha20

### PR DESCRIPTION
**Pull request by dhruv**
The current implementation of ChaCha20 [advances the block](https://github.com/bitcoin/bitcoin/blob/37633d2f61697fc719390767aae740ece978b074/src/crypto/chacha20.cpp#L310) even when the last pseudorandom block of 64 bytes is only partially used. This code buffers and reuses those pseudorandom bytes.

This improvement is relevant to the new design of BIP324 where:
- One instance of ChaCha20(for encrypting the length) uses only 3 pseudorandom bytes per message, so throwing away 61/64 bytes would be a 95% waste.
- There are efficiencies to be had from being able to send a vector of plaintexts to encrypt rather than creating a concatenated buffer, but that means the block cannot be advanced in between calls.

Unfortunately, this also means that our implementation is no longer identical to the reference implementation by Daniel J Bernstein (if I had to guess though, closer to his intention for a stream cipher), and the differential fuzz test against his implementation from #22704 no longer makes sense (if it's not against the pristine version, it does not serve the original purpose)

### CODE REVIEW

**Comment by dhruv** [comment]
@laanwj I don't have access to DJB but I tried to ask him upon your prompt: https://twitter.com/dhruv/status/1536354037552951296

**Comment by sipa** [comment]
@dhruv It looks intentional, but whatever it is, any protocols using it depend on it for better or worse - it's not like anyone can just switch out to the alternative without breaking. RFC8439 itself even seems to rely on it: it explains the fact that the last 32 bytes of the first input block (after the poly1305 key) are skipped.

**Comment by sipa** [comment]
Concept ACK on changing this; it much more closely matches the expectation of a stream cipher interface, especially as the specification does not say anything about skipping bytes.

That said, I think that wondering about the origin of this is mostly pointless apart from historical curiosity. It's an implementation issue, which seems to have leaked into the protocol definitions of things that have been built on top (incl. RFC8439 itself, which skips the second 32 byte of the chacha20 stream in the chacha20poly1305 cipher). Any protocol relying on it obviously can't just swap out this implementation as it'd be a protocol break, but other than that, either behavior can be implemented in function of either implementation (add a buffer on top in one direction, or introduce explicit skipping steps in the other).

**Comment by real-or-random** [comment]
It's just the API. See the comment starting with "Encryption/decryption of arbitrary length messages" in https://web.archive.org/web/20170119113129/http://www.ecrypt.eu.org/stream/svn/viewcvs.cgi/ecrypt/trunk/api/skel/ecrypt-sync.h?view=auto

**Comment by sipa** [comment]
> It's just the API. See the comment starting with "Encryption/decryption of arbitrary length messages" in https://web.archive.org/web/20170119113129/http://www.ecrypt.eu.org/stream/svn/viewcvs.cgi/ecrypt/trunk/api/skel/ecrypt-sync.h?view=auto

> but he is NOT allowed to make additional encryption calls once he
>  * has called ECRYPT_encrypt_bytes() (unless he starts a new message
>  * of course).

Right, that explains. It's still a bit bizarre that the function even bothers updating the position counter in the state then, but there's nothing wrong with it.

**Comment by dhruv** [comment]
@real-or-random Nice detective work! Thank you.

**Comment by laanwj** [comment]
> That said, I think that wondering about the origin of this is mostly pointless apart from historical curiosity. It's an implementation issue

Right. I'm happy that we found this issue before BIP324 was deployed. It would have been virtually impossible to make this optimization later.

**Comment by sipa** [comment]
I wrote a fuzz test for verifying that invoking the `ChaCha20::Crypto()` and/or `ChaCha20::Keystream()` function multiple times concatenated works identical to calling it once for the whole block: https://github.com/sipa/bitcoin/commits/pr25354. Feel free to steal/include/incorporate.

**Comment by dhruv** [comment]
> I wrote a fuzz test for verifying that invoking the ChaCha20::Crypto() and/or ChaCha20::Keystream() function multiple times concatenated works identical to calling it once for the whole block: https://github.com/sipa/bitcoin/commits/pr25354. Feel free to steal/include/incorporate.

Thanks, @sipa! Yes, I will incorporate it once I'm done refreshing all BIP324 PRs to the new spec in the next few days.

**Comment by dhruv** [comment]
Rebased with master for other branches in the project

File: src/Makefile.test.include [hunk 1]
```diff
@@ -246,7 +246,6 @@ test_fuzz_fuzz_SOURCES = \
  test/fuzz/crypto_chacha20.cpp \
  test/fuzz/crypto_chacha20_poly1305_aead.cpp \
  test/fuzz/crypto_common.cpp \
- test/fuzz/crypto_diff_fuzz_chacha20.cpp \
```
Comment (MarcoFalke) [inline src/Makefile.test.include:249]: nit:

```diff
diff --git a/test/sanitizer_suppressions/ubsan b/test/sanitizer_suppressions/ubsan
index 67ef512895..e698ffce53 100644
--- a/test/sanitizer_suppressions/ubsan
+++ b/test/sanitizer_suppressions/ubsan
@@ -19,7 +19,6 @@ unsigned-integer-overflow:bench/bench.h
 unsigned-integer-overflow:FuzzedDataProvider.h
 unsigned-integer-overflow:leveldb/
 unsigned-integer-overflow:minisketch/
-unsigned-integer-overflow:test/fuzz/crypto_diff_fuzz_chacha20.cpp
 implicit-integer-sign-change:*/include/boost/
 implicit-integer-sign-change:*/include/c++/
 implicit-integer-sign-change:*/new_allocator.h

**Review by theStack — COMMENTED** [review]
Concept ACK

In order to increase test coverage, I'd suggest to

* also add calls that consume more pseudorandom bytes than available from the buffered previous block; this tests that `to_use` is limited and that the the following two conditions are triggered (
https://github.com/dhruv/bitcoin/blob/70d2a4582cfbd063fc44238de9b445f546b5a514/src/crypto/chacha20.cpp#L123-L126):
```
            ...
            if (prev_block_start_pos >= 64) {
                prev_block_start_pos = 0;
            }
            if (bytes) continue;
            ...
```

* add a corresponding test for `ChaCha20::Crypt`

**Review by vincenzopalazzo — COMMENTED** [review]
Concept ACK

File: src/crypto/chacha20.cpp [hunk 2]
```diff
@@ -49,28 +49,37 @@ void ChaCha20::SetKey(const unsigned char* k, size_t keylen)
...
 {
     memset(input, 0, sizeof(input));
+    memset(prev_block_bytes, 0, sizeof(prev_block_bytes));
+    prev_block_start_pos = 0;
 }
 
 ChaCha20::ChaCha20(const unsigned char* k, size_t keylen)
 {
     SetKey(k, keylen);
+    prev_block_start_pos = 0;
```
Comment (stratospher) [inline src/crypto/chacha20.cpp:66]: 70d2a45:  could do `memset(prev_block_bytes, 0, sizeof(prev_block_bytes));` here too.

**Review by aureleoules — COMMENTED** [review]
## PR

`./src/bench/bench_bitcoin -filter='CHACHA20_1MB'`

|             ns/byte |              byte/s |    err% |        ins/byte |        cyc/byte |    IPC |       bra/byte |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                2.89 |      346,242,492.84 |    0.4% |           20.13 |            6.62 |  3.039 |           0.08 |    0.0% |      0.03 | `CHACHA20_1MB`


|             ns/byte |              byte/s |    err% |        ins/byte |        cyc/byte |    IPC |       bra/byte |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                2.88 |      346,669,699.46 |    0.6% |           20.13 |            6.59 |  3.052 |           0.08 |    0.0% |      0.03 | `CHACHA20_1MB`

## Master

`./src/bench/bench_bitcoin -filter='CHACHA20_1MB'`
|             ns/byte |              byte/s |    err% |        ins/byte |        cyc/byte |    IPC |       bra/byte |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                2.56 |      390,349,081.17 |    0.3% |           17.03 |            5.87 |  2.900 |           0.05 |    0.0% |      0.03 | `CHACHA20_1MB`

|             ns/byte |              byte/s |    err% |        ins/byte |        cyc/byte |    IPC |       bra/byte |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                2.57 |      389,658,756.50 |    0.4% |           17.03 |            5.89 |  2.892 |           0.05 |    0.0% |      0.03 | `CHACHA20_1MB`


is this implementation slower?

File: src/crypto/chacha20.cpp [hunk 3]
```diff
@@ -49,28 +49,37 @@ void ChaCha20::SetKey(const unsigned char* k, size_t keylen)
...
     input[15] = 0;
+
+    prev_block_start_pos = 0;
 }
 
 ChaCha20::ChaCha20()
 {
     memset(input, 0, sizeof(input));
+    memset(prev_block_bytes, 0, sizeof(prev_block_bytes));
+    prev_block_start_pos = 0;
```
Comment (aureleoules) [inline src/crypto/chacha20.cpp:60]: `prev_block_start_pos` already initialized in class

File: src/crypto/chacha20.cpp (hunk 2 above)
Comment (aureleoules) [inline src/crypto/chacha20.cpp:66]: `prev_block_start_pos` already initialized in class

**Comment by dhruv** [comment]
@aureleoules 

You're right, there's a slowdown. However, the existing implementation is incorrect and skipping blocks will be ~20x slower for the BIP324 specification. Here are my numbers:

master:
|             ns/byte |              byte/s |    err% |        ins/byte |        cyc/byte |    IPC |       bra/byte |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                1.25 |      799,288,251.65 |    0.1% |           17.33 |            4.74 |  3.654 |           0.05 |    0.0% |     11.02 | `CHACHA20_1MB`

#25354 (This PR):
|             ns/byte |              byte/s |    err% |        ins/byte |        cyc/byte |    IPC |       bra/byte |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                1.34 |      747,649,058.86 |    0.0% |           19.94 |            5.07 |  3.933 |           0.08 |    0.0% |     10.99 | `CHACHA20_1MB`

#26153 (Alternate implementation by @sipa):
|             ns/byte |              byte/s |    err% |        ins/byte |        cyc/byte |    IPC |       bra/byte |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                1.25 |      802,562,761.09 |    0.3% |           16.70 |            4.72 |  3.538 |           0.03 |    0.0% |     10.98 | `CHACHA20_1MB`

Since the implementation by @sipa meets the non-inferiority bar on the benchmarks (corroborated by @aureleoules' comment on #26153), I'm closing this PR and will shortly rebase all my branches for BIP324 to use #26153.

**Pull request closed without merging**